django/db/models/sql/query.py

```diff
@@ -934,8 +934,14 @@ def join(self, join, reuse=None, reuse_with_filtered_relation=False):
                 if (reuse is None or a in reuse) and j == join
             ]
         if reuse_aliases:
-            self.ref_alias(reuse_aliases[0])
-            return reuse_aliases[0]
+            if join.table_alias in reuse_aliases:
+                reuse_alias = join.table_alias
+            else:
+                # Reuse the most recent alias of the joined table, because
```

Comment:
Reuse the most recent alias of the joined table (a many-to-many relation may be joined multiple times).